### ISSUE TITLE
Ability to customize resource paths

### ISSUE THREAD

**Issue by ns-vpanfilov** [marmelab/react-admin#8723]
**Is your feature request related to a problem? Please describe.**

There are two problems that this request can solve:

**1. Provide ability to include multiple lists on the same page and have the ability to have lists with CREATE and UPDATE that work ouf of the box**

For example, if I am on a page `/self-service`, I want to display 2 lists and have update and create function. The way I do it today is by using base path manipulations, and by creating `AutoRegisterResource` custom component that registers and unregisters resources requested based on current location - it uses `useResourceDefinitionContext` to manipulate definitions, and also wrap modified definitions into a new context using `RsourceDefinitionContextProvider`

```js
// linked to `/self-service` custom route
// displays both tables; CREATE and EDIT functions work properly due to basename manipulation
// e.g. Edit for `domains` table will point to `/self-service/domains/:id`
const SelfServicePage = () => {

<BasenameContextProvider basename="/self-service/domains">
   <AutoRegisterResource name='domains' list={...} create={...} /
</BasenameContextProvider>
<BasenameContextProvider basename="/self-service/public_cloud_accounts">
   <AutoRegisterResource name='public_cloud_accounts' list={...} create={...} /
</BasenameContextProvider>
}
```
**2. Provide ability to include in tabbed interface that is automatically built using Children elements**
(e.g. I have a tabbed page `aws` with tabs `ec2` and `s3`, which are translated to `/aws/ec2/` and `/aws/s3` routes)

as in case of problem #1, I wrap components to `BasenameContextProvider` and `AutoRegisterResource`

Current challenges
1. Tab names must match resource names (this make it easy to make mistakes, and also creates undue complexities when trying to address the following 2 points)
2. Create nested tabs with resources is overly complex (requires heavy customization)
3. Create multiple lists in tabs is overly complex  (requires heavy customization)

**Describe the solution you'd like**

Add an optional parameter to `Resource` component that includes root path for Resource (rootPath). If the parameter is submitted, use it:
- When registering resource in CoreAdminUI.ts (instead of using `resource.props.name/*`, use `${resource.props.rootPath??resource.props.name}/*`
- Modify `useCreatePath()` to query resource definition using `useResourceDefinitionContext` to see if `rootPath` is set, and use its value instead of `${basename}/${resource}`

This way, the solution will not impact current users, and will give me the flexibility to map resources to different URLs easily.

**Comment by slax57** [comment]
Hi,
Thank you for your suggestion.

While this would be a nice feature to add, I'm afraid it would be quite time-consuming to implement fully, because there are several places in the code where we 'assume' the routes for the records to be in the form `/resource`, `/resource/id`, `/resource/id/show` and `/resource/create`, for example when redirecting.

I believe it should be doable to store the `rootPath` in the `ResourceDefinitionContext`, but then we would need to use this context in several places where we did not use it before (including some Enterprise packages).

Also, have you tried declaring your resources as for example `"self-service/domains"`? With some changes to the display labels in the menu and titles, and some tweaks in the dataProvider to map to the correct resource, maybe it would be an easier workaround?

Bottom line is, I'm not sure we are willing to spend too much time on this feature since it is not a common use-case, and it can be done in user-land (albeit admittedly in a cumbersome way).

@fzaninotto  or @djhi what do you think? Did you already encounter this use-case?

**Comment by fzaninotto** [comment]
I agree with @slax57. We've already discussed that in the past, and my opinion is that:

- it adds serious complexity in many places
- it's seldom used
- it 's doable in userland via duplicate resource names deduplicated at the dataProvider level. 
- 
So I'm -1 for this one.

**Comment by ns-vpanfilov** [comment]
i will declare resources as paths, then. thanks

**Comment by ns-vpanfilov** [comment]
I tried using duplicate resource names as @fzaninotto  suggested; it mostly worked; however there are some issues that make code complex and not very scalable. Would it be possible to allow users to override `useCreatePath` function, so we control how react-admin computes paths? This way, i will be able to create more than one table on the same route, and allow edit capabilities for each table:
- list /basepath - this page will have multiple resources
- edit: /basepath/resource/:id
- show: /basepath/resource/:id/show

**Comment by slax57** [comment]
Unfortunately no, I don't think this is possible.
I'm afraid you have no other choice but to manually set the paths for each component, or to create custom versions of all the components you need to use your own `useCreatePath`.

I'll close this issue since, as @fzaninotto explained, this is not a feature we are willing to integrate into RA's core.
